### PR TITLE
[FLINK-5008] Update quickstart documentation

### PR DESCRIPTION
**Pull request by NicoK**
This PR updates the outdated quickstart guides regarding IDE setup and the first example.

### CODE REVIEW

**Comment by greghogan** [comment]
I haven't yet looked at this PR in detail but it looks to have some much needed cleanup for the documentation.

Are we certain we need to abandon Eclipse as a viable IDE for Flink developers? The only thing worse than no support is a broken promise of support but have we tested later releases of the Scala IDE?

**Comment by NicoK** [comment]
I tried with several versions of Eclipse and Scala IDE, even with the one claimed to work. Unfortunately, I got none of them to work.

**Comment by thvasilo** [comment]
IIRC @vasia is using Eclipse to develop, maybe she can chime in.

**Comment by StephanEwen** [comment]
From my experience (having been an Eclipse user until one year ago), it is possible to develop a Flink Java Program in Eclipse. Developing Flink Scala programs is tricky.

Developing the Flink Code base in Eclipse is virtually impossible, because mixed Java/Scala projects work extremely bad.

Does it make sense to differentiate between developing the code base and developing applications?

**Comment by thvasilo** [comment]
Hello @NicoK , do you think you can include [FLINK-5087](https://issues.apache.org/jira/browse/FLINK-5087) in this PR, or should we create a new one?

**Comment by uce** [comment]
@thvasilo I think that should be handled in a separate PR. 

@StephanEwen Yes, I think this is a good differentiation to make in the docs. If it is not possible to write Flink programs with Eclipse, then we have a problem. I really hope that it is possible and we should check this. @NicoK Did you try to write a Flink program with Eclipse, for example using quick starts?

I think it's fine to drop the docs for Flink core dev with Eclipse if it is not possible to get it done in a straight forward manner. @vasia Do you have any input on this?

**Comment by vasia** [comment]
Sorry, no input from me regarding Eclipse. I've given up on it about a year ago ;)

**Comment by NicoK** [comment]
I'll look into writing Flink programs with Eclipse and update the documentation if needed

**Comment by NicoK** [comment]
Developing Flink programs still works with Eclipse (tested with Eclipse 4.6.1 and Scala IDE 4.4.1 for Scala 2.11). Alongside testing the quickstarts, I also updated them as promised and made a switch for showing unstable vs. stable documentation which differs slightly in setting up the example projects and getting Flink binaries. Please, have a look.

**Comment by NicoK** [comment]
I wasn't able to test the Scala SBT path though, so this may need some additional love by someone with a working SBT environment

File: docs/quickstart/java_api_quickstart.md [hunk 1]
```diff
@@ -46,39 +46,79 @@ Use one of the following commands to __create a project__:
...
+{% endif %}
     {% endhighlight %}
     </div>
 </div>
 
 ## Inspect Project
 
-There will be a new directory in your working directory. If you've used the _curl_ approach, the directory is called `quickstart`. Otherwise, it has the name of your artifactId.
+There will be a new directory in your working directory. If you've used
+the _curl_ approach, the directory is called `quickstart`. Otherwise,
```
Comment (fhueske) [inline docs/quickstart/java_api_quickstart.md]: "the _curl_ approach" -> "the _curl_ shortcut"?

File: docs/quickstart/java_api_quickstart.md [hunk 2]
```diff
@@ -46,39 +46,79 @@ Use one of the following commands to __create a project__:
...
+We recommend you __import this project into your IDE__ to develop and
+test it. If you use Eclipse, the [m2e plugin](http://www.eclipse.org/m2e/)
+allows to [import Maven projects](http://books.sonatype.com/m2eclipse-book/reference/creating-sect-importing-projects.html#fig-creating-import).
+Some Eclipse bundles include that plugin by default, others require you
+to install it manually. The IntelliJ IDE supports Maven projects out of
+the box.
 
 
-A note to Mac OS X users: The default JVM heapsize for Java is too small for Flink. You have to manually increase it. Choose "Run Configurations" -> Arguments and write into the "VM Arguments" box: "-Xmx800m" in Eclipse.
+*A note to Mac OS X users*: The default JVM heapsize for Java is too
```
Comment (fhueske) [inline docs/quickstart/java_api_quickstart.md]: Isn't the latest version called "macOS"?

File: docs/quickstart/java_api_quickstart.md [hunk 3]
```diff
@@ -46,39 +46,79 @@ Use one of the following commands to __create a project__:
...
+small for Flink. You have to manually increase it. In Eclipse, choose
```
Comment (fhueske) [inline docs/quickstart/java_api_quickstart.md]: Is that only the case for Eclipse users? 
If yes, it should be mentioned. If not, we should give instructions to increase the heapsize for IntelliJ as well.

File: docs/quickstart/java_api_quickstart.md [hunk 4]
```diff
@@ -46,39 +46,79 @@ Use one of the following commands to __create a project__:
...
+`Run Configurations -> Arguments` and write into the `VM Arguments`
+box: `-Xmx800m`.
 
 ## Build Project
 
-If you want to __build your project__, go to your project directory and issue the `mvn clean install -Pbuild-jar` command. You will __find a jar__ that runs on every Flink cluster in __target/your-artifact-id-{{ site.version }}.jar__. There is also a fat-jar,  __target/your-artifact-id-{{ site.version }}-flink-fat-jar.jar__. This
+If you want to __build your project__, go to your project directory and
+issue the `mvn clean install -Pbuild-jar` command. You will
+__find a jar__ that runs on every Flink cluster in
```
Comment (fhueske) [inline docs/quickstart/java_api_quickstart.md]: "every Flink cluster" -> "every Flink cluster with a compatible version"? Or rephrase the sentence.

File: docs/quickstart/java_api_quickstart.md [hunk 5]
```diff
@@ -46,39 +46,79 @@ Use one of the following commands to __create a project__:
...
+__target/original-your-artifact-id-your-version.jar__. There is also a
+fat-jar,  __target/your-artifact-id-your-version.jar__. This
 also contains all dependencies that get added to the maven project.
```
Comment (fhueske) [inline docs/quickstart/java_api_quickstart.md]: "maven" -> "Maven"

File: docs/quickstart/java_api_quickstart.md [hunk 6]
```diff
@@ -94,7 +134,9 @@ data 1
 is 1
 ~~~
 
-The following code shows the WordCount implementation from the Quickstart which processes some text lines with two operators (FlatMap and Reduce), and prints the resulting words and counts to std-out.
+The following code shows the `WordCount` implementation from the
+Quickstart which processes some text lines with two operators (FlatMap
+and Reduce), and prints the resulting words and counts to std-out.
```
Comment (fhueske) [inline docs/quickstart/java_api_quickstart.md]: The code does not show a `Reduce` but only `sum` which is executed as `Reduce`. Would be good to clarify this. Otherwise it might be confusing for readers.

File: docs/quickstart/run_example_quickstart.md [hunk 7]
```diff
@@ -90,23 +92,23 @@ use it in our program. Edit the `dependencies` section so that it looks like thi
     </dependency>
     <dependency>
         <groupId>org.apache.flink</groupId>
-        <artifactId>flink-streaming-java_2.10</artifactId>
+        <artifactId>flink-streaming-java_2.11</artifactId>
```
Comment (fhueske) [inline docs/quickstart/run_example_quickstart.md]: Why is this changed to Scala 2.11. Is this the default of the quickstarts?

File: docs/quickstart/scala_api_quickstart.md [hunk 8]
```diff
@@ -136,49 +136,82 @@ Use one of the following commands to __create a project__:
...
 {% endhighlight %}
     </div>
 </div>
 
 
 ### Inspect Project
 
-There will be a new directory in your working directory. If you've used the _curl_ approach, the directory is called `quickstart`. Otherwise, it has the name of your artifactId.
+There will be a new directory in your working directory. If you've used
+the _curl_ approach, the directory is called `quickstart`. Otherwise,
```
Comment (fhueske) [inline docs/quickstart/scala_api_quickstart.md]: "approach" -> "shortcut"?

File: docs/quickstart/scala_api_quickstart.md [hunk 9]
```diff
@@ -136,49 +136,82 @@ Use one of the following commands to __create a project__:
...
 
 ### Build Project
 
-If you want to __build your project__, go to your project directory and issue the `mvn clean package -Pbuild-jar` command. You will __find a jar__ that runs on every Flink cluster in __target/your-artifact-id-{{ site.version }}.jar__. There is also a fat-jar,  __target/your-artifact-id-{{ site.version }}-flink-fat-jar.jar__. This
+If you want to __build your project__, go to your project directory and
+issue the `mvn clean package -Pbuild-jar` command. You will
+__find a jar__ that runs on every Flink cluster in
+__target/original-your-artifact-id-your-version.jar__. There is also a
+fat-jar,  __target/your-artifact-id-your-version.jar__. This
 also contains all dependencies that get added to the maven project.
```
Comment (fhueske) [inline docs/quickstart/scala_api_quickstart.md]: maven -> Maven

File: docs/quickstart/scala_api_quickstart.md [hunk 10]
```diff
@@ -136,49 +136,82 @@ Use one of the following commands to __create a project__:
...
+The IntelliJ IDE supports Maven out of the box and offers a plugin for
+Scala development.
 
 
 ### Build Project
 
-If you want to __build your project__, go to your project directory and issue the `mvn clean package -Pbuild-jar` command. You will __find a jar__ that runs on every Flink cluster in __target/your-artifact-id-{{ site.version }}.jar__. There is also a fat-jar,  __target/your-artifact-id-{{ site.version }}-flink-fat-jar.jar__. This
+If you want to __build your project__, go to your project directory and
+issue the `mvn clean package -Pbuild-jar` command. You will
+__find a jar__ that runs on every Flink cluster in
```
Comment (fhueske) [inline docs/quickstart/scala_api_quickstart.md]: every Flink cluster (see above)

File: docs/quickstart/scala_api_quickstart.md [hunk 11]
```diff
@@ -194,7 +227,9 @@ data 1
 is 1
 ~~~
 
-The following code shows the WordCount implementation from the Quickstart which processes some text lines with two operators (FlatMap and Reduce), and prints the resulting words and counts to std-out.
+The following code shows the `WordCount` implementation from the
+Quickstart which processes some text lines with two operators (FlatMap
+and Reduce), and prints the resulting words and counts to std-out.
```
Comment (fhueske) [inline docs/quickstart/scala_api_quickstart.md]: sum / reduce: See comment above

File: docs/quickstart/setup_quickstart.md [hunk 12]
```diff
@@ -88,7 +111,9 @@ Now, we are going to run the [SocketWindowWordCount example](https://github.com/
     </div>
   </div>
 
-* Counts are printed to `stdout`. Monitor the JobManager's output file and write some text in `nc`:
+* Words are counted in time windows of 5 seconds (processing time, tumbling
+  windows) and are printed to `stdout`. Monitor the JobManager's output file
+  and write some text in `nc` (input is sent per line after hitting <RETURN>):
```
Comment (fhueske) [inline docs/quickstart/setup_quickstart.md]: "input is sent per line" -> "input is sent to Flink line by line"

File: docs/quickstart/setup_quickstart.md [hunk 13]
```diff
@@ -55,7 +69,10 @@ Check the __JobManager's web frontend__ at [http://localhost:8081](http://localh
 
 ## Run Example
 
-Now, we are going to run the [SocketWindowWordCount example](https://github.com/apache/flink/blob/master/flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/socket/SocketWindowWordCount.java) and read text from a socket and count the number of distinct words.
+Now, we are going to run the
+[SocketWindowWordCount example](https://github.com/apache/flink/blob/master/flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/socket/SocketWindowWordCount.java)
+and read text from a socket and count the number of distinct words per sliding
```
Comment (fhueske) [inline docs/quickstart/setup_quickstart.md]: Sliding window is changed to Tumbling window below

File: docs/quickstart/java_api_quickstart.md (hunk 2 above)
Comment (NicoK) [inline docs/quickstart/java_api_quickstart.md]: no idea but this probably applies to all Mac users or may not even be relevant to the new OS anyway so let's maybe keep it - btw: why is this only relevant to MacOS?

File: docs/quickstart/run_example_quickstart.md (hunk 7 above)
Comment (NicoK) [inline docs/quickstart/run_example_quickstart.md]: yes, for Flink-1.2 at least

File: docs/quickstart/java_api_quickstart.md (hunk 1 above)
Comment (NicoK) [inline docs/quickstart/java_api_quickstart.md]: shortcut doesn't sound right - at least to the user not knowing what is happening behind the script, you can consider the two options two different approaches of creating a quickstart project

Comment (fhueske) [inline docs/quickstart/java_api_quickstart.md]: Isn't a shortcut a faster way to do the same thing? The curled script calls the Maven command with predefined settings.

Comment (NicoK) [inline docs/quickstart/java_api_quickstart.md]: yes, that's what it does (internally) but wouldn't you say, a user not knowing these sees the two ways of creating the repo as two "approaches" and would be rather confused calling it a "shortcut"? - it could be made explicit but then we could argue why the script is needed in the first place... (the user will copy the command anyway)
however I did not change that and did not intend to

File: README.md [hunk 14]
```diff
@@ -104,25 +104,11 @@ Check out our [Setting up IntelliJ](https://github.com/apache/flink/blob/master/
...
-1. Download the Scala IDE (preferred) or install the plugin to Eclipse Kepler. See 
-   [How to setup Eclipse](https://github.com/apache/flink/blob/master/docs/internals/ide_setup.md#eclipse) for download links and instructions.
-2. Add the "macroparadise" compiler plugin to the Scala compiler.
-   Open "Window" -> "Preferences" -> "Scala" -> "Compiler" -> "Advanced" and put into the "Xplugin" field the path to
-   the *macroparadise* jar file (typically "/home/*-your-user-*/.m2/repository/org/scalamacros/paradise_2.10.4/2.0.1/paradise_2.10.4-2.0.1.jar").
-   Note: If you do not have the jar file, you probably did not run the command line build.
-3. Import the Flink Maven projects ("File" -> "Import" -> "Maven" -> "Existing Maven Projects") 
-4. During the import, Eclipse will ask to automatically install additional Maven build helper plugins.
-5. Close the "flink-java8" project. Since Eclipse Kepler does not support Java 8, you cannot develop this project.
+**NOTE:** From our experience, this setup does not work with Flink
```
Comment (NicoK) [inline README.md]: nice spotting this - changed that now

**Comment by fhueske** [comment]
Thanks for the update @NicoK!
Merging this PR
